tools/performance-server/shared/src/main/kotlin/org/jetbrains/influxdb/BenchmarkMeasurement.kt

```diff
@@ -0,0 +1,364 @@
+    }
+
+    override fun fromInfluxJson(data: JsonElement): List<Measurement> {
+        val points = mutableListOf<GoldenResultMeasurement>()
+        var columnsIndexes: Map<String, Int>
+        if (data is JsonObject) {
+            val results = data.getRequiredField("results") as JsonArray
+            results.map {
+                if (it is JsonObject) {
+                    val series = it.getRequiredField("series") as JsonArray
+                    series.map {
+                        if (it is JsonObject) {
+                            val columns = it.getRequiredField("columns") as JsonArray
+                            columnsIndexes = columns.mapIndexed{ index, it ->
+                                (it as JsonLiteral).unquoted() to index
+                            }.toMap()
+                            val valuesArrays = it.getRequiredField("values") as JsonArray
+                            valuesArrays.forEach {
+                                val values = it as JsonArray
+                                val name = (values[columnsIndexes["benchmark.name"]!!] as JsonLiteral).unquoted()
+                                val score = (values[columnsIndexes["benchmark.score"]!!] as JsonLiteral).double
+                                val metric = (values[columnsIndexes["benchmark.metric"]!!] as JsonLiteral).unquoted()
+                                points.add(GoldenResultMeasurement(name, metric, score))
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        return points
+    }
+}
+
+// Convert benchmarks measurements to standard benchmark report.
+fun List<BenchmarkMeasurement>.toReport(): BenchmarksReport? {
+    return this.firstOrNull()?. let { firstPoint ->
+        // Create common part of report.
+        val machine = Environment.Machine(firstPoint.envMachineCpu!!, firstPoint.envMachineOs!!)
+        val jdk = Environment.JDKInstance(firstPoint.envJDKVersion!!.value, firstPoint.envJDKVendor!!.value)
+        val environment = Environment(machine, jdk)
+        val backend = Compiler.Backend(Compiler.backendTypeFromString(firstPoint.kotlinBackendType!!)!!,
+                firstPoint.kotlinBackendVersion!!.value, firstPoint.kotlinBackendFlags!!)
+        val compiler = Compiler(backend, firstPoint.kotlinVersion!!.value)
+        val benchmarksList = map {
+            val metric = BenchmarkResult.metricFromString(it.benchmarkMetric!!)!!
+            BenchmarkResult(it.benchmarkName!! + metric.suffix, BenchmarkResult.statusFromString(it.benchmarkStatus!!.value)!!,
+                    it.benchmarkScore!!.value, metric,
+                    it.benchmarkRuntime!!.value, it.benchmarkRepeat!!.value, it.benchmarkWarmup!!.value)
+        }
+        BenchmarksReport(environment, benchmarksList, compiler)
+    }
+}
+
+// Database instance describing benchmark.
+class BenchmarkMeasurement : Measurement("benchmarks") {
```

Comment:
Why not make it data class and use predefined https://kotlinlang.org/docs/reference/data-classes.html?

Reply:
I don't understand what do you want to say? The idea was to have an instance which allows execute all queries. It isn't a data class.

Reply:
Then you could make state a data class and have pre-generated `copy` method, manual copying is call for trouble.

Reply:
I understand, but data class generates copy method that just call constructor. It isn't enough here, I need all other field be copied.
And I can't replace all fields in constructor, because they are created via delegated properties.